SIP: Nakamoto sBTC (#156)
stacksgov/sips: sips/sip-025/sip-025-sbtc.md

# Preamble

SIP Number: 022

Title: A Trustless 2-way Bitcoin Peg for Stacks

Authors:
* Aaron Blankstein <aaron@hiro.so>
* Mike Cohen <mjoecohen@gmail.com>
* Brice Dobry <brice@hiro.so>
* Marvin Janssen
* Friedger Müffke <mail@friedger.de>

> * Jesus Najera 

Reviewer: jnajera1917@gmail.com (thanks for flagging)

* Jude Nelson <jude@stacks.org>
* Don Park <donp@trustmachines.co>

> * Alie Slade <aslade@hiro.so>

Author: Thanks, will remove

* Andre Serrano <andre@resident.stacks.org>
* @Soju-Drinker 

> * Igor Sylvester <igor@trustmachines.co>

Author: Thanks, will remove.  Also, Don Park?

* Joey Yandle <joey@trustmachines.co> 

Consideration: Technical, Governance

Type: Consensus

Status: Draft

Created: 28 September 2023

License: CC0-1.0

Sign-off:

Layer: Consensus (hard fork)

Discussions-To: https://github.com/stacksgov/sips

# Abstract

> This SIP proposes the creation of a **trustless two-way peg** in which BTC (the

Author: Stacks is not intrinsically dependent on Bitcoin.  We use "burnchain" in all the other SIPs.

tokens of the current Stacks burnchain) can be manifested on Stacks as a
SIP-010-compliant fungible token, called sBTC, and later redeemed for the same
number of BTC.  This proposal redefines the act of Stacking to be the act of
maintaining a _peg wallet_, which is used to hold BTC and process sBTC
redemptions.  Stackers are collectively responsible for ensuring that all
liquid sBTC are backed by an equal number of BTC, and that redemptions are
processed in a timely manner.  As a reward, they continue to receive PoX
payouts.

# Introduction

The lack of a stateful smart contract system on Bitcoin necessitates the
construction of systems in which a representation of locked Bitcoin is traded
within a separate smart-contract-powered blockchain.  These systems aim to

> provide a "2-way peg" between the representation of the locked BTC ("wrapped

Reviewer 2: [nit] consistently use two-way or 2-way, pick one. Also, why is "2-way peg" in quotes?

Author: It's a term of art being introduced in this SIP.  Will emphasize instead of quote.  Also, 2-way peg it is.

BTC") and the BTC itself.  At a high level, these systems seek to provide two
primitive operations:

> * "Peg-in": a BTC holder rids themselves of their BTC, and in doing so,

Reviewer 2: [nit] I recall some discussion on using deposit/withdraw instead of peg-in/peg-out?

Author: Sure, I like those terms better.

requests.

To support this, sBTC offers two modes of operation:
the Normal mode, and the Recovery mode.  In **Normal mode**, the sBTC asset is
backed 1:1 by BTC sent to a wallet controlled by a large fraction of Stackers,
as measured by the fraction of locked STX they represent.  Each time BTC is
sent to this wallet (a peg-in operation), an equal number of sBTC are
automatically minted to an address of the sender's choosing.  Stackers respond
to peg-out requests by sending BTC to the requester's Bitcoin address, thereby
maintaining the peg.  The act of responding to peg-out requests automatically
destroys the requester's equivalent amount of sBTC on the Stacks chain.

If the Normal mode encounters a liveness failure for any reason (including loss
of BTC), the system transitions to a Recovery mode until enough Stackers come
back online.  In **Recovery mode**, a fraction of the PoX payouts are
redirected to peg-out requests such that eventually, all outstanding requests
will be fulfilled, even if the Stackers _never_ come back online.  While
considerably slower than Normal mode, the design of Recovery mode ensures all

> sBTC can be redeemed so long as the Stacks blockchain and PoX are online.

Reviewer 2: [nit] "PoX are online" sounds weird (PoX not a separate system or service). Maybe "so long as the Stacks blockchain is operational and PoX is active" is clearer?

Author: PoX is a subsystem within the blockchain that can be disabled while the blockchain is online.

This proposal, if ratified, would be executed by at least two hard forks:  one
to implement Normal mode, and a follow-on hard fork to implement Recovery mode.
This SIP is designed to leverage features proposed by SIP-021, and can only
activate if SIP-021 is ratified beforehand or concurrently with this SIP.

> This proposal describes Normal mode.  Recovery mode will be fully described in a

Reviewer 2: Hmm, will recovery mode be implemented though? Or is the plan to launch with just normal mode?

Author: It's not part of this SIP, so the reference implementation need not implement it for this SIP to activate.

future SIP, and will have a separate activation procedure.  The Recovery mode
SIP cannot activate without this SIP first being ratified.

# Specification

The proposed SIP-021 upgrade would do away with Stacks forks, which vastly
simplifies the sBTC design.  Importantly, miners and Stackers cannot "roll back"
sBTC operations, meaning that peg-in and peg-out operations follow at-most-once
semantics.  This lack of "fork risk" means that the system can support
arbitrarily large amounts of BTC pegged-in and pegged-out, assuming that at
least 70% of Stackers by weight are trustworthy.

The sBTC system state is materialized in a boot contract called `.sbtc`.  This
contract, like `.pox-4`, is managed in part by the node itself.  In particular,
the node writes sBTC state from Bitcoin into the data space of `.sbtc` (such as
via a private function that only the node can call), so that it can be accessed
by smart contracts in Stacks.

The following sections describe the Normal mode of operation.  Recovery mode is

> left for future work, but is described in a separate section after Normal mode.

Reviewer 2: earlier in the doc you say recovery mode will be described in a future SIP

Author: A synopsis is provided.  I'll make that clear here.

in an `OP_RETURN` output (Figure 1).  Because this is a Stacks-on-Bitcoin
transaction, a miner in a subsequent tenure is required by the consensus rules
to process it as part of producing the first block in its tenure.  The Stacks
node records the existence of this peg-in operation in the data space of the
`.sbtc` smart contract, where it is marked as a _pending_ operation that
Stackers must act upon.  Namely, the relevant parts of the transaction are
extracted and translated in to Clarity values and made available via internal data maps.

Stackers, via their signer daemons, already subscribe to the Stacks blockchain
for events.  The act of writing new sBTC peg-in state to the `.sbtc` contract
generates such an event -- an event that alerts the signer daemons to begin a
signing round to act on the peg-in.

In the happy path, at least 70% of Stackers agree that the peg-in transaction is
valid, and they generate a contract-call to the `.sbtc` contract to
_authorize_ the peg-in by minting the equivalent number of sBTC tokens to
an account address of the user's choice.  Crucially, the transaction is signed 

> collectively by Stackers via a WSTS signing round -- it is authenticated via a

Reviewer 2: [nit] first mention of WSTS, can you include a link or pointer for further reading?

Schnorr proof and the Stacker's aggregate public key.  This way, _only_ the
Stackers can authorize the peg-in, and only via a 70%+ majority vote.

Once stackers generate the authorization transaction, they make it available to
miners via their StackerDB instance, and require the current miner to include it
in the next block they produce.  If the miner does not do so, then the Stackers

> will not sign the block; this all but compells the miner to include the

Reviewer 2: [typo] s/compells/compels

start of the next reward cycle.  If the peg-out succeeds, then they receive BTC from
the Stackers (minus a BTC transaction fee), and their sBTC gets burnt._

When Stackers see the peg-out request, and if they are able to act on it (i.e.
the wallet has enough UTXOs free), they execute a WSTS signing round to
generate a Bitcoin transaction from the peg wallet's UTXOs which pays out to the
user's requested address.  Because all pegged-in BTC is spendable by the current
set of Stackers via an aggregate Schnorr signature, Stackers simply create the
scriptSigs for sufficiently many UTXOs to fund a fulfillment UTXO for the user.

Because fulfilling the redemption request occurs through WSTS, there is a
possibility of a quorum failure.  As such, the Stackers may instead vote to
abstain or even to refuse the peg-out request.  This is explained in the
"Failure Modes" section.

The BTC redemption transaction contains the following fields:

* An `OP_RETURN` identifying the peg-out request in the `.sbtc` contract

* A new UTXO that redeems the user's sBTC for BTC

* A change UTXO that consolidates the remaining unused BTC from the consumed
  wallet UTXOs

> Once the Stackers have created the BTC redemption transaction, they both send it

Reviewer 2: [nit] "they both sending it ..." was confusing to read. Removing "both" or moving it after send seems clearer? something like "they broadcast it to both a Bitcoin peer and to the Stacks network"

Author: Sure -- will remove

`.sbtc` contract's data space.  This is done for failure recovery: if the
Bitcoin transaction does not confirm quickly, the Stackers can try again (see
below).  Also, the existence of the signed Bitcoin transaction in the `.sbtc`
data space serves as proof that the Stackers handled the peg-out request in a
timely manner, even if it was not confirmed on Bitcoin as quickly as desired.
When the `.sbtc` contract processes the copy of the BTC redemption transaction,
it marks the UTXOs it consumes as "tentatively used" so that a
concurrently-generated BTC redemption transaction dees not also use them.

Once the BTC redemption transaction confirms on Bitcoin, a Stacks miner
processes it as a Stacks-on-Bitcoin transaction.  The act of processing this
transaction is the act of calling into the `.sbtc` contract to mark the pending
sBTC peg-out as fulfilled.  This burns the user's locked sBTC, and marks the
peg-out operation as complete.  It also updates the `.sbtc` transaction's
knowledge of the wallet's UTXO set by clearing the UTXOs consumed by the
fulfillment, and adding the change UTXO as a new UTXO that the wallet controls.

### Peg-Out Congestion

> A key responsibility of the `.sbtc` wallet is to keep track of which Bitcoin

Reviewer 2: [nit] what's the distinction between `.sbtc` wallet and "peg wallet"? If they're the same, considering using one variant for consistency

Author: `.sbtc` is a smart contract.  The peg wallet is the set of UTXOs that can be spent by the Stackers to process withdrawals.

Once all of the outstanding UTXOs have been consumed by peg-transfers, the new
Stackers begin handling peg-ins and peg-outs for the next reward cycle.

### Recovery of BTC Fees

The transfer is complete once all of the old wallet's UTXOs (both unused and
pending) are consumed by peg-transfer transactions.  However,
the total BTC in the peg wallet will be slightly less than the
total amount of sBTC.  This is because (1) the old Stackers will have had to pay
transaction fees to send the peg hand-off transactions, and (2) Stackers
may have had to RBF a stuck BTC redeem transaction.

These losses due to BTC fees are distributed across sBTC accounts lazily when the user next
pegs in BTC, transfers the sBTC, or attempts to redeem it.  The `.sbtc` contract keeps track
of the total supply of sBTC, the balance of each account, and the number of
missing satoshis, so it has enough information to make this happen.  When the user attempts
to do any of the above with sBTC, the `.sbtc` contract will burn an amount of sBTC equal to
the following:

```
min(1, (sbtc_balance * missing_satoshis) / sbtc_supply)
```

In other words, the sBTC holder will pay a pro-rata share of the missing BTC

> fees, or it will pay a single satoshi of it.  An sBTC holder pays at most one

Reviewer 2: [typo] s/one/once

the reward cycle, or the old Stackers cannot hand off all of the BTC to the new
Stackers by the start of the new reward cycle. The only way to thaw the system
is for the new Stackers to eventually complete their vote and the old Stackers
send over the BTC before the next PoX anchor block is mined.  If the system
remains frozen this way, and the next PoX anchor block is mined, then the Stacks
blockchain halts.

### System Recovery from Frozen Peg

In the event that the BTC gets "stuck" and Stackers are unable to access it, the
system will support a variation of a peg-in transaction called a _donation_.
The only difference between a peg-in and a donation is that the donation does
not materialize sBTC.  It merely gives Stackers some BTC with which to fulfill
peg-outs.  This feature is meant to enable Stackers to recover lost BTC and
unfreeze the system, should the need ever arise.
 
## Initial Capital Limits

This SIP proposes that the total amount of sBTC accepted into this system is
capped by a limit in the `.sbtc` contract.  The `.sbtc` contract will refuse to
admit a peg-in if it would exceed this limit.  This limit can be increased (but

> not decreased) by Stacker vote, and it would start at a default value of 100 BTC.

Reviewer 2: What's the process to change this limit by stacker vote?

Reviewer 2: Any further context on the 100 BTC limit? As in why not 1000 or 10000 or 10?

Author: It's up to the implementation of `.sbtc`.  This SIP does not call for a specific means.

100 BTC is just over 1% of 70% of the TVL in PoX as of today.  We could safely bump it to about 7,500 BTC before the wallet would become worth more than 70% of the locked STX.  I'm open to alternatives and compelling arguments; my rationale for a small number is to give us some time to see how well it behaves on mainnet without incurring too much risk up front.  I'm not really worried about 70% of the stackers stealing the BTC; I'm worried about someone finding and exploiting a bug that lets them clean out the sBTC peg wallet without needing the stackers' permission.

The reason for the limit is to allow the system to be stress-tested while
minimizing risk to users in the early days.  Once Stackers are comfortable with
maintaining the peg and running signer daemons to manage the sBTC wallet,

> they can increase this limit.  They would remove it by simply setting the limit

Reviewer 2: Would this require an updated `.sbtc` boot contract (another hard-fork)? If not, what's the mechanism by which this would happen?

Author: No; `.sbtc` would provide the means of allowing stackers to vote on the maximum number of sBTC that are allowed to be in circulation.  Stackers can later vote to increase this limit without a hard fork.

# Preamble

SIP Number: 022

> Title: A Trustless 2-way Bitcoin Peg for Stacks

Reviewer 3: ```suggestion
Title: A Trust-minimized 2-way Bitcoin Peg for Stacks
```

The term "trust-minimized" seems to be preferred for describing the trust model in a succinct way, as discussed [here](https://twitter.com/muneeb/status/1712884857523638459).

cc @muneeb-ali

Status: Draft

Created: 28 September 2023

License: CC0-1.0

Sign-off:

Layer: Consensus (hard fork)

Discussions-To: https://github.com/stacksgov/sips

# Abstract

This SIP proposes the creation of a **trustless 2-way peg** in which BTC (the
tokens of the current Stacks burnchain) can be manifested on Stacks as a
SIP-010-compliant fungible token, called sBTC, and later redeemed for the same
number of BTC.  This proposal redefines the act of Stacking to be the act of
maintaining an _sBTC wallet_, which is used to hold BTC and process sBTC
redemptions.  Stackers are collectively responsible for ensuring that all
liquid sBTC are backed by an equal number of BTC, and that redemptions are
processed in a timely manner.  As a reward, they continue to receive PoX
payouts.

# Introduction

The lack of a stateful smart contract system on Bitcoin necessitates the
construction of systems in which a representation of locked Bitcoin is traded
within a separate smart-contract-powered blockchain.  These systems aim to
provide a _2-way peg_ between the representation of the locked BTC ("wrapped
BTC") and the BTC itself.  At a high level, these systems seek to provide two
primitive operations:

> * "Deposit": a BTC holder rids themselves of their BTC, and in doing so,

Reviewer: While I agree in the "deposit" & "withdraw" nomenclature, it'd be helpful to mention that the former can be thought of as "peg-in" & the latter as a "peg-out" in this definition section.

Especially since the term "peg" is used throughout.

Author: We're trying not to use the word "peg"

Reviewer 4: We use the word "peg" in practice, so while it isn't **technically** the best word, its a practical word. I'm generally pro-practical.

holder, _even if every miner is malicious_.  This is achieved by ensuring that
all Stacking and wallet maintenance operations materialize in all Stacks forks.

* **If a Byzantine fault-tolerant (BFT) majority of miners are honest, then sBTC
remains live**.  All withdraw requests are eventually fulfilled, _even if every
Stacker is malicious_.  This is because Stackers do not have a say in block
production, and Stackers that do not fulfill withdraws lose their PoX-driven BTC
income from miners.

* **Withdraws of arbitrary amounts of BTC are fulfilled in
a fixed amount of time on the happy path** if both Stackers and miners operate with a BFT honest
majority.  If they do not, then withdraws of arbitrary amounts of BTC 
are fulfilled eventually by _redirecting_ Stackers' PoX payouts to fund withdraw
requests.

To support this, sBTC offers two modes of operation:
the Normal mode, and the Recovery mode.  In **Normal mode**, the sBTC asset is
backed 1:1 by BTC sent to a wallet controlled by a large fraction of Stackers,
as measured by the fraction of locked STX they represent.  Each time BTC is
sent to this wallet (a deposit operation), an equal number of sBTC are

> automatically minted to an address of the sender's choosing.  Stackers respond

Reviewer: More accurate to say "automatically inserted for processing" vs. "automatically minted"? 

The current form makes it sound faultless & guaranteed.

by smart contracts in Stacks.

The following sections describe the Normal mode of operation.  Recovery mode is
left for future work, but is summarized in a separate section after Normal mode.
This SIP can activate without Recovery mode.

## Deposit

A deposit operation is initiated as a Bitcoin transaction with a special payload
in an `OP_RETURN` output (Figure 1).  Because this is a Stacks-on-Bitcoin
transaction, a miner in a subsequent tenure is required by the consensus rules
to process it as part of producing the first block in its tenure.  The Stacks
node records the existence of this deposit operation in the data space of the
`.sbtc` smart contract, where it is marked as a _pending_ operation that
Stackers must act upon.  Namely, the relevant parts of the transaction are
extracted and translated in to Clarity values and made available via internal data maps.

Stackers, via their signer daemons, already subscribe to the Stacks blockchain
for events.  The act of writing new sBTC deposit state to the `.sbtc` contract
generates such an event -- an event that alerts the signer daemons to begin a
signing round to act on the deposit.

> In the happy path, at least 70% of Stackers agree that the deposit transaction is

Reviewer 5: is this 70% weighted by amount locked ?

Author: Yes

Reviewer 4: Answered.

valid, and they generate a contract-call to the `.sbtc` contract to
_authorize_ the deposit by minting the equivalent number of sBTC tokens to
an account address of the user's choice.  Crucially, the transaction is signed 
collectively by Stackers via a WSTS signing round -- it is authenticated via a
Schnorr proof and the Stacker's aggregate public key.  This way, _only_ the
Stackers can authorize the deposit, and only via a 70%+ majority vote.

Once stackers generate the authorization transaction, they make it available to

> miners via their StackerDB instance, and require the current miner to include it

Reviewer: A bit unclear on what this means? Doesn't StackerDB only hold list tuples of signer slots?

in the next block they produce.  If the miner does not do so, then the Stackers
will not sign the block; this all but compels the miner to include the
transaction.

The user must wait until their deposit transaction confirms on Bitcoin before the
sBTC tokens materialize.  This is required because Stacks miners only process
pending Stacks-on-Bitcoin transactions for Bitcoin blocks up to but _excluding_
the Bitcoin block in which their `block-commit` landed.

The act of fulfilling a deposit with an affirmative vote is effectively a promise

> made by Stackers to users to ensure that the sBTC minted wll remain fully backed

Reviewer 6: minted **will** remain

Reviewer 4: Answered.

by the BTC the user submitted.  This BTC may only be spent to fulfill withdraws,

> or to process a wallet hand-off at the end of a reward cycle.  Either way, the

Reviewer: Might it be worth defining reward cycle earlier as a tentative two-week period? If someone is reading without stacking/PoX experience it might not be clear how long a "reward cycle" is.

Reviewer 4: Not sure if that is a safe assumption given not all of us understand it either, should add it to the glossary in the next PR. Making a note.

user's BTC will be spent by Stackers before they can reclaim it via the spending
fulfillment script.

![Figure 1](./001-sip-025-figure-1.png)

_Figure 1: Protocol diagram of an sBTC deposit.  The user submits a Bitcoin
transaction, and eventually, the equivalent number of sBTC tokens materialize in
the account of the user's choosing_.

### Deposit Bitcoin Transaction

The deposit Bitcoin transaction contains the following data:

> * An `OP_RETURN` data output, which contains the account address to which sBTC

Reviewer 6: Where should the specification for this go? Currently, the Alpha Romeo DR 0.1 has some specs: https://github.com/stacks-network/sbtc/blob/f4130c63ec5680921617ac4b910d72ffb4072eb5/sbtc-core/src/operations/op_return/deposit.rs#L4

> * An `OP_RETURN` data output, which contains the account address to which sBTC

Reviewer 5: Is using OP_DROP to embed the deposit data excluded in this spec ?

Reviewer 4: Added in subsequent comment. We discuss this in detail in https://github.com/stacks-network/sbtc/discussions/315

  should be minted

> * A payment to the sBTC sBTC wallet maintained by the current Stackers.  This

Reviewer: "sBTC sBTC" typo?

>   payment UTXO is the hash of a script that has two spending fulfillment

Reviewer: Could we specific / link the exact formatting of this script? I believe I've seen some nice visuals around.

> conditions:  either the Stackers spend the UTXO before the end of the next

Author: Yes, and yes

> conditions:  either the Stackers spend the UTXO before the end of the next

Reviewer 4: Answered.

reward cycle, or the sender can spend the UTXO once the end of the next reward
cycle has passed.

The payment must be above a `.sbtc`-defined threshold, in satoshis.  This

> _minimum deposit_ value is used to prevent users from spamming the system with

Reviewer: What is the initial MDV?

Author: Was hoping you could tell me :)  I was thinking the `.sbtc` contract would stipulate it via a Stacker mandate, but we'll want to pick an initial value.

, * The Bitcoin redeem script for the user's payment.  The Stackers must verify
  that the script contains the aforementioned spending fulfillment conditions,
_and no others_.  The script is disclosed on the Stacks chain in order to (1)
allow 3rd parties to authenticate the script and (2) check the Stackers' work in
order to hold them accountable.

On successful execution, the sBTC will be minted to the user-specified Stacks
account, and the redeem script will be stored in the `.sbtc` contract for
posterity.  The `.sbtc` contract does not attempt to parse or validate the
redeem script, but it does keep track of all of the UTXOs that the wallet
has assumed control over via "yes" votes.  In addition, it stores the voting
results and bitmaps for posterity.

There is no transaction fee for the Stacker authorization transaction.  Stackers
instead require the miner to include it for free in their blocks by withholding
block signatures until it is included.

### Failure Modes and Recovery

There are several places where a deposit can fail.  First, the Stackers may fail
to complete a signing round that achieves the 70% threshold.  One of the
following outcomes will happen:

> * The Stackers produce a deposit authorization with at least 70% of the STX

Reviewer 5: should this read 30% or more of the STX voting against.. ?

Author: One implies the other

Reviewer 4: Resolved.

the Stackers.  This would prevent the Stackers' vote from being recorded in
`.sbtc`.  If this happens, then Stackers refuse to sign any blocks until they
see a block which contains the as-yet-unmined authorization transaction(s).  If
need be, Stackers themselves can spend some BTC to become miners in order to
push their transactions through.

### Transaction Fees

The user pays the transaction fee for their sBTC deposit transaction.  Stackers
do not pay a STX transaction fee, but instead require miners to include their
deposit authorization transaction in the next block they sign.

### Two-phase Deposit

Users may not be able to send a Bitcoin transaction with an `OP_RETURN` output.
For example, this is difficult to do or is unsupported in some Bitcoin wallets,
especially custodial wallets.  To support these users, a 3rd party could simply
send deposit transaction on their behalf.  The user would simply pay the 3rd
party the requisite Bitcoin to deposit, and the 3rd party would generate and send
the deposit transaction for them.

## Withdraw

Withdrawing is the act of redeeming sBTC for BTC (Figure 2).

> To redeem sBTC for BTC, the user first sends a Stacks trnsaction to `.sbtc`

Reviewer 6: The current Alpha Romeo 0.1 developer release has implemented the withdrawal as a btc transaction. I suggest to add the btc withdrawal tx as a second option?

Author: Sure, this can be done.

> To redeem sBTC for BTC, the user first sends a Stacks trnsaction to `.sbtc`

Reviewer 5: typo - Stacks **trnsaction**

Reviewer 4: Done.

  system freezes until enough Stackers can come online.  This has penalties,
which are described in a following section.

## BTC Transfer

A new set of Stackers assumes responsibility for the sBTC wallet during the
prepare phase for their reward cycle.  Due to SIP-021, the Stacks blockchain
never forks, which means that the new set of Stackers is known immediately after
the last non-prepare-phase Stacks block is mined (since this block is guaranteed
to be the PoX anchor block).

Once the PoX anchor block is known, the miner _immediately_ calculates the new
PoX reward set (including the list of Stackers and their signing keys), and
writes them into the `.pox-4` data space.  This would be required by the
consensus rules -- the block would not be valid if the miner did not do this.
Today, the node does this calculation
at the end of the prepare phase, and only writes the state into the sortition
database's MARF; this proposal calls for doing this at the start of the prepare
phase, and writing this information into both places.

### Deciding the New sBTC Wallet

Once the new set of Stackers is known, they have 99 Bitcoin blocks to carry out

> a WSTS DKG round amongst themselves and write the aggregate public key and BTC

Author: DKG == distributed key generation

> a WSTS DKG round amongst themselves and write the aggregate public key and BTC

Reviewer 4: Added in glossary.

least the sBTC users will eventually be made whole in the event of a
catastrophe.

In Recovery mode, sBTC deposits would cease and all pending sBTC deposits would be
cancelled.  Withdraws would be fulfilled in first-in first-out order from
redirected PoX payouts.  Once a withdraw request has been
satisfied -- i.e. the requester's BTC address has
received _at least_ as many satoshis as reqeusted from the redirected PoX
payouts -- then the withdraw request completes and the next requester's withdraw
is processed.

Once the Stacks blockchain supports Recovery Mode, the system would continue
operating even if the system was frozen for many reward cycles.  The blockchain
would not halt; instead, it would continue its operation to fulfill all withdraw
requests while the BTC is seemingly inaccessible to Stackers.  PoX payouts and
STX unlocks would only resume once the current set of Stackers control
enough BTC to back each sBTC token, and are able to fulfill pending sBTC
withdraws.

The system would transition back to Normal mode once all pending withdraw
transactions have been fulfilled, and once the current set of Stackers possess
enough BTC to cover all outstanding sBTC tokens.

> # Related Work

Reviewer 3: Should add a section about [Nomic](https://gist.github.com/mappum/da11e37f4e90891642a52621594d03f6), which is the most similar related work imo.

Author: cc @andreserrano

There are several other systems which implement a wrapped BTC asset on a
different chain.  The key difference between most of them and sBTC is that the
nodes which control the BTC wallet use a closed-membership protocol to decide
which parties can enter and leave the set.  In detail:

> ## wBTC 
> 
> This is a closed membership system.  It is made up of
> 50+ merchants and custodians with keys to the wBTC
> multi-sig contract on Ethereum. End users purchase wBTC directly from
> authorized merchants, and often the experience is fast.  However, wBTC deposits and
> withdrawals can only be performed by authorized merchants and custodians. 

Reviewer 3: ```suggestion
## WBTC 

This is a closed membership system.  It is made up of
50+ merchants and custodians with keys to the WBTC
multi-sig contract on Ethereum. End users purchase WBTC directly from
authorized merchants, and often the experience is fast.  However, WBTC deposits and
withdrawals can only be performed by authorized merchants and custodians. 
```

It is stylized `WBTC` as seen on their website: https://wbtc.network/

Might be worth noting that although the merchants manage issuance and redemption, all BTC backing WBTC is held by a single company (BitGo).

Author: cc @andreserrano

## RBTC

This is a closed membership system.  Nine organizations supporting the Powpeg,
and there is no direct control of Bitcoin multi-sig private keys. Peg operations
settle to Bitcoin via merge mining on RSK side-chain.

## BTC.b

This is a closed membership system.  It is operated by eight wardens (relayers)
and a private codebase, secured by an SGX enclave. The SGX enclave only
processes transfers when it receives approvals from at least 6 of the 8 Warden nodes.

## LBTC

> This is a closed membership system.  Over 50 members of Liquid Federation manage multi-sig

Reviewer 3: The Liquid multisig is an 11-of-15 multisig. [Source](https://help.blockstream.com/hc/en-us/articles/900001408623-How-does-Liquid-Bitcoin-L-BTC-work-)

> contract.  Only federation members can process withdraws.

Reviewer 3: ```suggestion
contract.  Only federation members can process withdrawals.
```

typo fix

Author: cc @andreserrano

addition to their block-signing responsibilities in SIP-021.

In order for this SIP to activate, the following criteria must be met by the set
of Stacked STX:

* At least 80 million Stacked STX must vote _at all_ to activate this SIP.  This
  number is chosen because it is more than double the amount of STX locked by
the largest Stacker at the time of this writing (reward cycle 69).

* Of the Stacked STX that vote, at least 80% of them must vote "yes."

The act of not voting is the act of siding with the outcome, whatever it may be.
We believe that these thresholds are sufficient to demonstrate interest from
Stackers -- Stacks users who have a long-term interest in the Stacks
blockchain's succesful operation -- in performing this upgrade.

### How To Vote

If a user is Stacking, then their STX can be used to vote in one of two ways,
depending on whether or not they are solo-stacking or stacking through a
delegate.

The user must be Stacking in any cycle up to and including cycle XXX.  Their vote
contribution will be the number of STX they have locked.

#### Solo Stacking

The user must send a minimal amount of BTC from their PoX reward address to one

> of the following Bitcoin addresses:

Reviewer 3: Are the BTC sent to these addresses considered unspendable?

Author: Well, if they are spendable, you're more than welcome to claim them yourself ;)

Reviewer 3: @jcnelson I will take that as a "yes"? The problem with these unspendable "burn addresses" is that they pollute the UTXO set with UTXOs that can never be spent, so they just bloat the UTXO set forever for no good reason. We have a better solution for burning BTC, it is by specifying an OP_RETURN output in a transaction and assigning an amount of BTC to be spent in this output. The BTC is then provably burnt but it does not bloat the UTXO set. (This is the difference between "the simple way" and "the meaningful way" defined [here](https://medium.com/@alcio/how-to-destroy-bitcoins-255bb6f2142e))

Reviewer 4: @jcnelson is this part still relevant with the intent of putting the burned BTC from the prepare phase into the wallet's warchest?

* Jesus Najera <jnajera1917@gmail.com>
* Joey Yandle <joey@trustmachines.co>
* Jude Nelson <jude@stacks.org>
* Marvin Janssen <marvin@ryder.id>
* Mike Cohen <mjoecohen@gmail.com>
* Sayak Chatterjee <sayak@trustmachines.co>
* Stjepan Golemac <stjepan@trustmachines.co>
* Tycho Onnasch <tycho@tychoonnasch.com>

Consideration: Technical, Governance

Type: Consensus

Status: Draft

Created: 28 September 2023

License: CC0-1.0

Sign-off:

Layer: Consensus (hard fork)

Discussions-To: https://github.com/stacksgov/sips

# Abstract

This SIP proposes the creation of a **trust-minimized 2-way peg** in which BTC (the
tokens of the current Stacks burnchain) can be manifested on Stacks as a
SIP-010-compliant fungible token, called sBTC, and later redeemed for the same
number of BTC.  This proposal redefines the act of Stacking to be the act of
maintaining an _sBTC wallet_, which is used to hold BTC and process sBTC
redemptions.  Stackers are collectively responsible for ensuring that all
liquid sBTC are backed by an equal number of BTC, and that redemptions are
processed in a timely manner.  As a reward, they continue to receive PoX
payouts.

# Introduction

## Glossary

> | Term | Definition |

Reviewer 4: Add `Reward cycle` with a link to SIP-007 as requested in https://github.com/stacksgov/sips/pull/156#discussion_r1361225487

backed 1:1 by BTC sent to a wallet controlled by a large fraction of Stackers,
as measured by the fraction of locked STX they represent.  Each time BTC is
sent to this wallet (a deposit operation), an equal number of sBTC are
automatically minted to an address of the sender's choosing.  Stackers respond
to withdraw requests by sending BTC to the requester's Bitcoin address, thereby
maintaining the peg.  The act of responding to withdraw requests automatically
destroys the requester's equivalent amount of sBTC on the Stacks chain.

If the Normal mode encounters a liveness failure for any reason (including loss
of BTC), the system transitions to a Recovery mode until enough Stackers come
back online.

In **Recovery mode**, a fraction of the PoX payouts are
redirected to withdraw requests such that eventually, all outstanding requests
will be fulfilled, even if the Stackers _never_ come back online.  While
considerably slower than Normal mode, the design of Recovery mode ensures all
sBTC can be redeemed so long as the Stacks blockchain and PoX are online.

## Scope

This proposal only describes Normal mode, and leaves Recovery mode to be described
in a future SIP.

> This proposal, if ratified, would be executed by at least two hard forks:  one

Reviewer 4: Add a specification `Maintenance` mode to encompass the need for manual recovery and is addressed later in the document with the details in https://github.com/stacks-network/sbtc/discussions/346 as mitigation strategies.

`Maintenance` mode is the temporary placeholder for `Recovery` mode prior to implementation.

activate if it is ratified beforehand or concurrently with this SIP.

The Recovery mode SIP cannot activate without this SIP first being ratified.

# Specification

The proposed [SIP-021][SIP-021-LINK] upgrade would do away with Stacks forks, which vastly
simplifies the sBTC design over what would be needed for a fork tolerant system.
Importantly, miners and Stackers cannot "roll back" sBTC operations, meaning that deposit
and withdraw operations follow at-most-once semantics. This lack of "fork risk" means
that the system can support arbitrarily large amounts of BTC deposited and withdrawn,
assuming that at least 70% of Stackers by weight are trustworthy.

The sBTC system state is materialized in a boot contract called `.sbtc`.  This
contract, like `.pox-4`, is managed in part by the node itself.  In particular,
the node writes sBTC state from Bitcoin into the data space of `.sbtc` (such as
via a private function that only the node can call), so that it can be accessed
by smart contracts in Stacks.

The following sections describe the Normal mode of operation.  Recovery mode is
left for future work, but is summarized in a separate section after Normal mode.

> ## Deposit

Reviewer 4: @FriendsFerdinand I wrote this section with a mostly complete understanding of OP_DROP but am missing some details. Please suggest adjustments when you can.

I've based this section on these discussions:
1. https://github.com/stacks-network/sbtc/discussions/314
2. https://github.com/stacks-network/sbtc/discussions/315

Schnorr proof and the Stacker's aggregate public key.  This way, _only_ the
Stackers can authorize the deposit, and only via a 70%+ majority vote.

Once stackers generate the authorization transaction, they make it available to
miners via their StackerDB instance, and require the current miner to include it
in the next block they produce.  If the miner does not do so, then the Stackers
will not sign the block; this all but compels the miner to include the
transaction.

The user must wait until their deposit transaction confirms on Bitcoin before the
sBTC tokens materialize.  This is required because Stacks miners only process
pending Stacks-on-Bitcoin transactions for Bitcoin blocks up to but _excluding_
the Bitcoin block in which their `block-commit` landed.

The act of fulfilling a deposit with an affirmative vote is effectively a promise
made by Stackers to users to ensure that the sBTC minted wll remain fully backed
by the BTC the user submitted.  This BTC may only be spent to fulfill withdrawals,
or to process a DKG wallet hand-off.  Either way, the user's BTC will be spent by
Stackers before they can reclaim it via the spending fulfillment script.

> ***TODO: https://github.com/stacks-network/sbtc/issues/349: Create Deposit Procedure Diagram***

Reviewer 4: @FriendsFerdinand can you address this ticket?

https://github.com/stacks-network/sbtc/issues/349

### Deposit Bitcoin Transaction

The deposit Bitcoin transaction sends Bitcoin to the generated address with the
following script spending paths:

1. The sbtc payload is embedded in the script that sends the Bitcoin to the signer's pubkey
    - `<sbtc_payload> DROP <signers_pubkey> CHECKSIG`
2. The Refund path
    - `<block-height> CHECKLOCKTIMEVERIFY DROP <user_pubkey> CHECKSIG`

### Deposit Clarity Contract Call

The `.sbtc` clarity contract will need to contain a function that, when called,
does the following:

1. Verifies that the Bitcoin transaction is included in the bitcoin block of a specified height
2. Verifies the inclusion of the 2 script paths tap leaves in the output (the root is generated from the 2 tap leaves)
3. Records the existence of the Transaction ID, the amount sent and the time at which it was committed in the `.sbtc` data space (the height at which this Bitcoin transaction was included, not at which it was verified in the Clarity contract)

> #### `.sbtc` Contract Function

Reviewer 4: Do we need these details in the SIP?

copy of the in-flight Bitcoin transaction in the `.sbtc` contract, so the
`.sbtc` contract can update its view of the available wallet UTXOs.  This would
be achieved by a contract-call to `.sbtc` to carry out the replacement.  As
before, Stackers would require miners to include this transaction in their
blocks in order for the blocks to be signed.

The act of replacing an in-flight transaction in `.sbtc` is also the act of
updating its knowledge of how much _extra_ BTC was spent out of the sBTC wallet
to cover the additional transaction fee.  This is important for the hand-off
(see below).

### Transaction Fees

Bitcoin transaction fees are unavoidable.  This proposal calls for users to pay
the BTC transaction fee for both their deposits and withdraws.  As such, there
exists a minimum BTC redemption amount:  a given redemption request, minus the
fees required to produce a BTC redemption transaction, must be worth more than
the Bitcoin dust minimum (5,500 satoshis).  If it is not, then the Stackers NACK
the withdraw request.

Stackers set the required satoshis/vbyte transaction fee rate in the `.sbtc`
contract data space, so users can discover it at any time.  Stackers update this

> fee rate via a collectively-signed contract-call transactionto `.sbtc`, which

Reviewer 7: "transactionto" typo

produce a deposit UTXO, which contains the hash of a redeem script; the redeem
script determines the sBTC wallet's address.

![Figure 3](./003-sip-025-figure-3.png)

_Figure 3: Protocol diagram of a BTC-transfer.  Once the PoX anchor block is known,
the new Stackers proceed to execute a WSTS DKG and post the new aggregate public key and
wallet redeem script to Stacks.  The old Stackers help them do this by requiring miners
to include their votes in their blocks as a prerequisite for signing them.  Once the
new public key and wallet redeem script have reached 70% support, the old Stackers send
one or more Bitcoin transactions which consume all of the current wallet's UTXOs (including those
for in-flight withdraws) and create new ones for the new sBTC wallet.  Once all old UTXOs
are consumed, then the hand-off completes.  The act of processing a BTC-transfer transaction
updates the .sbtc contract's wallet UTXO tracker per the consensus rules._

The Stackers write the aggregate public key and BTC redeem script
through an on-chain vote.  The new Stackers submit their votes as zero-fee Stacks
transactions, which they share with existing Stackers so they can compel miners

> include them in blocks (and should this fail for any reason, Stackers can also send

Reviewer 7: include -> to include

the following:

```
min(1, (sbtc_balance * missing_satoshis) / sbtc_supply)
```

In other words, the sBTC holder will pay a pro-rata share of the missing BTC
fees, or it will pay a single satoshi of it.  An sBTC holder pays at most once
per reward cycle.  Eventually, as long as there is
sBTC activity on-chain, the number of sBTC in circulation will be made equal to
the number of satoshis in the sBTC wallet.  If this fails to happen by the end of
the reward cycle, then the sBTC system freezes and a Stacker may need to
manually pay BTC into the sBTC wallet to recover the fee (see "System Recovery
from Frozen Wallet").

### Restarting In-Flight Transactions

If a BTC-transfer happens while there are in-flight redemption transactions, then the old
Stackers must double-spend the in-flight redemption transactions.  The `.sbtc`
contract identifies which in-flight redemption transactions would be affected by
observing which UTXOs the BTC-transfer transaction consumes.  Each affected in-flight
redemption transaction would be marked as cancelled, and the user's sBTC
unlocks.  The user will need to try the request again.

> ### Transfer Fees

Reviewer 4: I chose the word `peg` for a lot of this section for lack of a better word, though I understand that we want to avoid that term. [comment link](https://github.com/stacksgov/sips/pull/156/files#r1362485277)

1. Why
2. Can someone suggest better wording?

Reviewer 3: How about `bridge`?

The peg wallet will need to be transferred every reward cycle, which is an unbounded number of
transfers with no tie to the numebr deposit and withdrawal requests. This means there is
no way to charge a fee that, at the moment of deposit or withdrawal, is
guaranteed to adequately pay for the peg transfer fees for the duration of the pegged
Bitcoin's tenure.

To pay for peg wallet transfer fees without dipping into user funds, this SIP proposes
the following mitigation strategies.

1. During the prepare phase, the traditionally burned bitcoin will instead be placed into the peg wallet. See [SIP-001][SIP-001-LINK]

> 2. The protocol will maintain a warchest that can receive donations and be siphoned from if the wallet budget for transfer fees runs dry.

Reviewer 4: This should be altered to make it clear that the "warchest" is just a concept that maps to "the amount of BTC in the peg wallet that's not accounted for in the sBTC system"

Reviewer 3: How about a term that doesn't have "war" connotations? Like "treasury" or "surplus" or something like that.

BTC to the new Stackers.  This act will cancel all pending sBTC withdraw
requests.

* On BTC-transfer, the system can freeze in one of two ways:  the new Stackers
cannot vote on a new aggregate public key and redeem script before the end of
the reward cycle, or the old Stackers cannot hand off all of the BTC to the new
Stackers by the start of the new reward cycle. The only way to thaw the system
is for the new Stackers to eventually complete their vote and the old Stackers
send over the BTC before the next PoX anchor block is mined.  If the system
remains frozen this way, and the next PoX anchor block is mined, then the Stacks
blockchain halts.

### System Recovery from Frozen Wallet

In the event that the BTC gets "stuck" and Stackers are unable to access it, the
system will support a variation of a deposit transaction called a _donation_.
The only difference between a deposit and a donation is that the donation does
not materialize sBTC.  It merely gives Stackers some BTC with which to fulfill
withdraws.  This feature is meant to enable Stackers to recover lost BTC and
unfreeze the system, should the need ever arise.

## Capital Limits

> **No capital limit will be imposed on the system.** Instead, the protocol will rely

Reviewer 4: @andreserrano This will likely stay the same through future iterations, so please make any alterations.
